injection/sharedmain/main.go

```diff
@@ -245,3 +287,89 @@ func flush(logger *zap.SugaredLogger) {
+// by calling log.Fatalf.
+func MemStatsOrDie(ctx context.Context) {
+	msp := metrics.NewMemStatsAll()
+	msp.Start(ctx, 30*time.Second)
+
+	if err := view.Register(msp.DefaultViews()...); err != nil {
+		log.Fatalf("Error exporting go memstats view: %v", err)
+	}
+}
+
+// SetupLoggerOrDie sets up the logger using the config from the given context
+// and returns a logger and atomic level, or dies by calling log.Fatalf.
+func SetupLoggerOrDie(ctx context.Context, component string) (*zap.SugaredLogger, zap.AtomicLevel) {
+	loggingConfig, err := GetLoggingConfig(ctx)
+	if err != nil {
+		log.Fatalf("Error reading/parsing logging configuration: %v", err)
+	}
+	logger, atomicLevel := logging.NewLoggerFromConfig(loggingConfig, component)
+	ctx = logging.WithLogger(ctx, logger)
+
+	return logger, atomicLevel
+}
+
+// SetupConfigMapWatchOrDie establishes a watch of the configmaps in the system
+// namespace that are labeled to be watched or dies by calling log.Fatalf.
+func SetupConfigMapWatchOrDie(ctx context.Context, logger *zap.SugaredLogger) *configmap.InformedWatcher {
+	kc := kubeclient.Get(ctx)
+	// Create ConfigMaps watcher with optional label-based filter.
+	var cmLabelReqs []labels.Requirement
+	if cmLabel := system.ResourceLabel(); cmLabel != "" {
+		req, err := configmap.FilterConfigByLabelExists(cmLabel)
+		if err != nil {
+			logger.With(zap.Error(err)).Fatalf("Failed to generate requirement for label %q")
+		}
+		logger.Infof("Setting up ConfigMap watcher with label selector %q", req)
+		cmLabelReqs = append(cmLabelReqs, *req)
+	}
+	// TODO(mattmoor): This should itself take a context and be injection-based.
+	cmw := configmap.NewInformedWatcher(kc, system.Namespace(), cmLabelReqs...)
+	return cmw
+}
+
+// WatchLoggingConfigOrDie establishes a watch of the logging config or dies by
+// calling log.Fatalf.
+func WatchLoggingConfigOrDie(ctx context.Context, cmw *configmap.InformedWatcher, logger *zap.SugaredLogger, atomicLevel zap.AtomicLevel, component string) {
+	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(logging.ConfigMapName(),
+		metav1.GetOptions{}); err == nil {
+		cmw.Watch(logging.ConfigMapName(), logging.UpdateLevelFromConfigMap(logger, atomicLevel, component))
+	} else if !apierrors.IsNotFound(err) {
+		logger.With(zap.Error(err)).Fatalf("Error reading ConfigMap %q", logging.ConfigMapName())
+	}
+}
+
+// WatchObservabilityConfigOrDie establishes a watch of the logging config or
+// dies by calling log.Fatalf.
+func WatchObservabilityConfigOrDie(ctx context.Context, cmw *configmap.InformedWatcher, profilingHandler *profiling.Handler, logger *zap.SugaredLogger, component string) {
+	if _, err := kubeclient.Get(ctx).CoreV1().ConfigMaps(system.Namespace()).Get(metrics.ConfigMapName(),
+		metav1.GetOptions{}); err == nil {
+		cmw.Watch(metrics.ConfigMapName(),
+			metrics.UpdateExporterFromConfigMap(component, logger),
+			profilingHandler.UpdateFromConfigMap)
+	} else if !apierrors.IsNotFound(err) {
```

Comment:
ditto here, what happens if it fails for another reason?

Reply:
Same here.